docs/proposal/2022-12-23-rbac.md

```diff
@@ -0,0 +1,399 @@
+
+  ```go
+  &discordgo.MessageCreate{
+    Message: &discordgo.Message{
+      // ...
+      Member: &discordgo.Member{
+        // ...
+        Roles: []string{
+          "976789858670497805", // <-- this is what we can use
+        },
+      },
+    },
+  }
+  ```
+
+  Based on that we can query for a full role object to get its name.
+
+  ```go
+  b.api.AddHandler(func(s *discordgo.Session, m *discordgo.MessageCreate) {
+      msg := discordMessage{
+        Event: m,
+      }
+      roles, _ := b.api.GuildRoles(m.GuildID)
+      // manually filter role by ID - unfortunately there's just list capability - couldn't find get
+          // we can cache the roles as they are server-wide (or guild-wide, depends what term is used)
+      // get role.Name - more: https://discord.com/developers/docs/topics/permissions#role-object
+  ```
+
+MS Teams:
+
+- We need to do additional API calls to get the user name email. Firstly, we need to [retrieve user ID and/or channel ID](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/channel-and-group-conversations?tabs=json#retrieve-mentions), then [get user details](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/get-teams-context?tabs=json#get-single-member-details).
+- There are no user groups or roles in MS Teams we can use.
+
+### Configure mapping
+
+User creates Roles and ClusterRoles, as well as the RoleBindings and ClusterRoleBindings. For example:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: read-only
+subjects:
+  - kind: Group # group binding
+    name: developers
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User # user binding
+    name: foo@example.com
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
+  kind: ClusterRole
+  name: read-only
+  apiGroup: rbac.authorization.k8s.io
+```
+
+Next, Botkube needs to be aware of such configuration. In order to achieve this, we'll extend our execution configuration. We would support multiple types of mapping:
+
+- user-email based (allows per-user access rights configuration)
+- channel-based (allows per-channel access rights configuration)
+- user-group based (for supported platforms we'll use user groups or roles; for MS Teams, we'll fallback to channel-based mapping)
+- static (allows to impersonate a given user or group every time a given executor is executed regardless the context)
+
+> **NOTE:** Username and groups prefix could be useful in the clusters where OIDC is configured. For example, see [this](https://fusionauth.io/blog/2022/02/24/rbac-with-kubernetes-fusionauth) blog post.
+
+```yaml
+executors:
+  'kubectl-read-only':
+    botkube/kubectl:
+      enabled: true
+      # New field "context" which will be used by Botkube Default Executor
+      context:
+        kubeconfig: &kubeconfigCtx
+
+            # Option 1: user impersonation using only email address
+            mapping:
```

Comment:
- What about moving this to same indentation with `kubeconfig`? Not all of the executors may need kubeconfig, so that we can use exactly same mapping strategy for all plugins.
- `mapping` seems very broad term, what about using `rbac` or something like that?
- `context` seems good, especially isolating rbac stuff from other plugin parameters 👌

Reply:
1. Do you mean to get rid of the second level, that is instead of `executors[name][pluginName].context.kubeconfig.nmapping` have `executors[name][pluginName].context.kubeconfig` or (`executors[name][pluginName].context.rbac`)? I think this mapping will be used only for Kubernetes cluster 🤔 Do you see any possibilities to reuse it in a different way? I would like to understand other cases, as I thought about this RBAC as something 100% related to Kubernetes, which could be wrong thinking indeed. BTW I'm totally fine to decrease the nesting level and/or rename the properties 👍 
3. You're right
4. 👌

Reply:
I meant something like this

```
context:
  kubeconfig: ....
  mapping: ...
``` 

Remember Mateusz showed sample plugin in one of his demos, github plugin that creates github issues. I would like to say, "the user with email foo@example.com in channel #dev can create github issues". If this is totally out of context of this proposal, then freely ignore 1. item.

Reply:
Ah, alright, so you mean to unify the RBAC mapping config across all the plugins? Sure, I think that's great idea, as long as we are sure they will be unified 👍 Then the `context.kubeconfig` would be removed, as `mapping` is currently the only property there, so that's why there won't be a need to have something like this 👍  So yeah, let's remmove `context.kubeconfig` and move `mapping` to the top of the context, and rename it to e.g. `rbac` or something similar 👍

Reply:
I have more or less the same feedback as @huseyinbabal. The current structure is a bit too complex, but proposed simplification is ok with me 👍 

Probably later it would be good to pass the `userInfo` object to each executor.

Reply:
> Probably later it would be good to pass the userInfo object to each executor.

Do you mean allowing cluster admin to define whole userInfo, including `uid` and `extra`? Not sure if there will be demand for that, but yes, this config is extendable for future, so 👍 At some point we can even extract `extra` and `uid` from user profiles if necessary, but probably that would need some argumentation and concrete use cases 😄

Reply:
Ah, I think I understand now - so you mean getting k8s `userInfo` using  `SelfSubjectReview` API to get data about a given user and pass it to the plugin, correct? Or do you mean some other arbitrary object with data got from config and comm platform APIs? Either way, that's totally doable, but let's see first if that will be useful 👍I'll add a new sentence:

> This path may be used by the plugin to access Kubernetes cluster in a restricted way. Later we can introduce more fields to the context, such as user details.